crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -269,48 +271,244 @@ impl ClassInfoConstraintFunction {
+/// This is a disjunction (OR) of conjunctions (AND) of constraints.
+/// The DNF representation allows us to properly track `TypeGuard` constraints
+/// through boolean operations.
+///
+/// For example:
+/// - `f(x) and g(x)` where f returns `TypeIs[A]` and g returns `TypeGuard[B]`
+///   => `[Conjunction { constraint: A, typeguard: Some(B) }]`
+///   => evaluates to `B` (`TypeGuard` clobbers)
+///
+/// - `f(x) or g(x)` where f returns `TypeIs[A]` and g returns `TypeGuard[B]`
+///   => `[Conjunction { constraint: A, typeguard: None }, Conjunction { constraint: object, typeguard: Some(B) }]`
+///   => evaluates to `A | B`
+#[derive(Hash, PartialEq, Debug, Eq, Clone)]
+struct NarrowingConstraint<'db> {
+    /// Disjunctions of conjunctions (DNF)
+    disjuncts: SmallVec<[Conjunction<'db>; 4]>,
+}
+
+impl get_size2::GetSize for NarrowingConstraint<'_> {}
+
+impl<'db> NarrowingConstraint<'db> {
+    /// Create a constraint from a regular (non-`TypeGuard`) type
+    fn regular(constraint: Type<'db>) -> Self {
+        Self {
+            disjuncts: smallvec![Conjunction::regular(constraint)],
+        }
+    }
+
+    /// Create a constraint from a `TypeGuard` type
+    fn typeguard(constraint: Type<'db>) -> Self {
+        Self {
+            disjuncts: smallvec![Conjunction::typeguard(constraint)],
+        }
+    }
+
+    /// Evaluate the type this effectively constrains to
+    ///
+    /// Forgets whether each constraint originated from a `TypeGuard` or not
+    fn evaluate_type_constraint(self, db: &'db dyn Db) -> Type<'db> {
+        UnionType::from_elements(
+            db,
+            self.disjuncts
+                .into_iter()
+                .map(|disjunct| Conjunction::evaluate_type_constraint(disjunct, db)),
+        )
+    }
+}
+
+impl<'db> From<Type<'db>> for NarrowingConstraint<'db> {
+    fn from(constraint: Type<'db>) -> Self {
+        Self::regular(constraint)
+    }
+}
+
+/// Internal representation of constraints with DNF structure for tracking `TypeGuard` semantics.
+///
+/// This is a newtype wrapper around `FxHashMap<ScopedPlaceId, NarrowingConstraint<'db>>` that
+/// provides methods for working with constraints during boolean operation evaluation.
+#[derive(Clone, Debug, Default)]
+struct InternalConstraints<'db> {
+    constraints: FxHashMap<ScopedPlaceId, NarrowingConstraint<'db>>,
+}
+
+impl<'db> InternalConstraints<'db> {
+    /// Insert a regular (non-`TypeGuard`) constraint for a place
+    fn insert_regular(&mut self, place: ScopedPlaceId, ty: Type<'db>) {
+        self.constraints
+            .insert(place, NarrowingConstraint::regular(ty));
+    }
+
+    /// Convert internal constraints to public constraints by evaluating each DNF constraint to a Type
+    fn evaluate_type_constraints(self, db: &'db dyn Db) -> NarrowingConstraints<'db> {
+        self.constraints
+            .into_iter()
+            .map(|(place, constraint)| (place, constraint.evaluate_type_constraint(db)))
+            .collect()
+    }
```

Comment:
I don't think this method can exist in a correct implementation, there is no correct way to do this translation without access to the prior type for each place.